File: backend/app/models/investment.py
"""Investment model."""
import uuid
from datetime import datetime, date
from decimal import Decimal
from sqlalchemy import Column, String, DateTime, Date, Boolean, Enum, Numeric, ForeignKey
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship
import enum
from app.database import Base


class InvestmentType(str, enum.Enum):
    """Investment type enumeration."""
    PPF = "PPF"
    ELSS = "ELSS"
    NPS = "NPS"
    FD = "FD"
    MUTUAL_FUND = "MUTUAL_FUND"
    STOCK = "STOCK"
    REAL_ESTATE = "REAL_ESTATE"
    GOLD = "GOLD"
    CRYPTO = "CRYPTO"
    LIC = "LIC"
    SUKANYA_SAMRIDDHI = "SUKANYA_SAMRIDDHI"
    NSC = "NSC"
    OTHER = "OTHER"


class TaxSection(str, enum.Enum):
    """Tax section enumeration."""
    SEC_80C = "80C"
    SEC_80CCC = "80CCC"
    SEC_80CCD = "80CCD"
    SEC_80D = "80D"
    SEC_80E = "80E"
    SEC_80G = "80G"
    NONE = "NONE"


class AssetLiabilityCategory(str, enum.Enum):
    """Asset/Liability category enumeration."""
    ASSET = "ASSET"
    LIABILITY = "LIABILITY"


class Investment(Base):
    """Investment model for tracking investments and assets."""
    
    __tablename__ = "investments"
    
    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    user_id = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=False)
    name = Column(String, nullable=False)
    investment_type = Column(Enum(InvestmentType), nullable=False)
    amount_invested = Column(Numeric(15, 2), nullable=False)
    current_value = Column(Numeric(15, 2), nullable=True)
    annual_return_pct = Column(Numeric(5, 2), nullable=True)
    start_date = Column(Date, nullable=False)
    maturity_date = Column(Date, nullable=True)
    is_tax_saving = Column(Boolean, default=False, nullable=False)
    tax_section = Column(Enum(TaxSection), default=TaxSection.NONE, nullable=False)
    folio_number = Column(String, nullable=True)
    ticker_symbol = Column(String, nullable=True)
    is_active = Column(Boolean, default=True, nullable=False)
    rich_dad_category = Column(Enum(AssetLiabilityCategory), default=AssetLiabilityCategory.ASSET, nullable=False)
    passive_income_amount = Column(Numeric(15, 2), default=0, nullable=False)
    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False)
    
    # Relationships
    user = relationship("User", back_populates="investments")
    tax_deductions = relationship("TaxDeduction", back_populates="investment")

File: frontend/lib/models/investment.dart
/// Investment model
enum InvestmentType {
  PPF,
  ELSS,
  NPS,
  FD,
  MUTUAL_FUND,
  STOCK,
  REAL_ESTATE,
  GOLD,
  CRYPTO,
  LIC,
  SUKANYA_SAMRIDDHI,
  NSC,
  OTHER
}

enum TaxSection { 
  SEC_80C, 
  SEC_80CCC, 
  SEC_80CCD, 
  SEC_80D, 
  SEC_80E, 
  SEC_80G, 
  NONE 
}

enum AssetLiabilityCategory { ASSET, LIABILITY }

class Investment {
  final String id;
  final String userId;
  final String name;
  final InvestmentType investmentType;
  final double amountInvested;
  final double? currentValue;
  final double? annualReturnPct;
  final DateTime startDate;
  final DateTime? maturityDate;
  final bool isTaxSaving;
  final TaxSection taxSection;
  final String? folioNumber;
  final String? tickerSymbol;
  final bool isActive;
  final AssetLiabilityCategory richDadCategory;
  final double passiveIncomeAmount;
  final DateTime createdAt;
  final DateTime updatedAt;

  Investment({
    required this.id,
    required this.userId,
    required this.name,
    required this.investmentType,
    required this.amountInvested,
    this.currentValue,
    this.annualReturnPct,
    required this.startDate,
    this.maturityDate,
    this.isTaxSaving = false,
    this.taxSection = TaxSection.NONE,
    this.folioNumber,
    this.tickerSymbol,
    this.isActive = true,
    this.richDadCategory = AssetLiabilityCategory.ASSET,
    this.passiveIncomeAmount = 0,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Investment.fromJson(Map<String, dynamic> json) {
    return Investment(
      id: json['id'],
      userId: json['user_id'],
      name: json['name'],
      investmentType: InvestmentType.values.byName(json['investment_type']),
      amountInvested: (json['amount_invested'] as num).toDouble(),
      currentValue: json['current_value'] != null 
          ? (json['current_value'] as num).toDouble() 
          : null,
      annualReturnPct: json['annual_return_pct'] != null
          ? (json['annual_return_pct'] as num).toDouble()
          : null,
      startDate: DateTime.parse(json['start_date']),
      maturityDate: json['maturity_date'] != null
          ? DateTime.parse(json['maturity_date'])
          : null,
      isTaxSaving: json['is_tax_saving'] ?? false,
      taxSection: TaxSection.values.byName(json['tax_section'].replaceAll('SEC_', '')),
      folioNumber: json['folio_number'],
      tickerSymbol: json['ticker_symbol'],
      isActive: json['is_active'] ?? true,
      richDadCategory: AssetLiabilityCategory.values.byName(json['rich_dad_category']),
      passiveIncomeAmount: (json['passive_income_amount'] as num).toDouble(),
      createdAt: DateTime.parse(json['created_at']),
      updatedAt: DateTime.parse(json['updated_at']),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'name': name,
      'investment_type': investmentType.name,
      'amount_invested': amountInvested,
      'current_value': currentValue,
      'annual_return_pct': annualReturnPct,
      'start_date': startDate.toIso8601String().split('T')[0],
      'maturity_date': maturityDate?.toIso8601String().split('T')[0],
      'is_tax_saving': isTaxSaving,
      'tax_section': taxSection.name,
      'folio_number': folioNumber,
      'ticker_symbol': tickerSymbol,
      'is_active': isActive,
      'rich_dad_category': richDadCategory.name,
      'passive_income_amount': passiveIncomeAmount,
    };
  }
}
